runtime-transaction/src/instruction_details.rs

```diff
@@ -0,0 +1,273 @@
+use {
+    crate::{builtin_instruction_details::*, compute_budget_instruction_details::*},
+    solana_compute_budget::compute_budget_limits::*,
+    solana_sdk::{
+        instruction::CompiledInstruction,
+        pubkey::Pubkey,
+        transaction::{Result, TransactionError},
+    },
+    std::num::NonZeroU32,
+};
+
+/// Information about instructions gathered after scan over transaction;
+/// These are "raw" information that suitable for cache and reuse.
+#[cfg_attr(test, derive(Eq, PartialEq))]
+#[derive(Default, Debug)]
+pub struct InstructionDetails {
+    builtin_instruction_details: BuiltinInstructionDetails,
+    compute_budget_instruction_details: ComputeBudgetInstructionDetails,
+}
+
+impl InstructionDetails {
+    pub fn try_from<'a>(
+        instructions: impl Iterator<Item = (&'a Pubkey, &'a CompiledInstruction)>,
+    ) -> Result<Self> {
+        let mut compute_budget_instruction_details = ComputeBudgetInstructionDetails::default();
+        let mut builtin_instruction_details = BuiltinInstructionDetails::default();
+
+        for (i, (program_id, instruction)) in instructions.enumerate() {
```

Comment:
This approach of going instruction by instruction and comparing each program id is not very efficient. We should probably use a deduped program id list to find the compute budget program id, filter relevant instructions for processing, and then build the struct of the tx requested compute budget parameters.

Reply:
This is an interesting idea to explore. So the idea is to collect as many instruction related information in one iteration, so we don't have to scan tx many times. For the immediate needs, `InstructionDetails` currently collects info about compute-budget and builtin instructions. We could filter just builtin instructions first. I am also planning to add pre-compiled info into `instructionDetails` (such as num_signatures), that might also be able to fit into same schema. I'd like to think about it more, maybe can do this in follow up PRs? As for this PR, it can continue keep currently behavior.

Reply:
Please don't resolve conversations that aren't resolved

Reply:
Can you try benching this approach as well?

Reply:
> Please don't resolve conversations that aren't resolved

I closed it cause I thought this conversation has merged into above one, there I added filter and bench results. Sound like you have different idea to try here. Can you explain more? By "use a deduped program id list", I am assuming you are referring to compute-budget programs specifically, but we are not just processing compute budget anymore, also want to accumulate all builtin's cost.

Reply:
I think the idea is that a `program_id_index` may appear multiple times. We do not need to resolve it to builtin/compute-budget each time we see that index - only the first time.